go/runtime/log/log.go

```diff
@@ -0,0 +1,126 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/nxadm/tail"
+)
+
+// Log is a log file handle.
+//
+// It implements `io.Writer` so it can be used as the log handler and it also provides methods for
+// streaming the content of the log.
+type Log struct {
+	mu      sync.Mutex
+	file    *os.File
+	closeCh chan struct{}
+
+	maxSize     int
+	currentSize int
+}
+
+func newLog(fn string, maxSize int) (*Log, error) {
+	file, err := os.OpenFile(fn, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Log{
+		file:    file,
+		closeCh: make(chan struct{}),
+		maxSize: maxSize,
+	}, nil
+}
+
+// Write writes the given buffer into the log file.
+//
+// In case the total written size exceeds the configured maximum, the log file is rotated.
+func (l *Log) Write(p []byte) (int, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	n, err := l.file.Write(p)
+	if err != nil {
+		return 0, err
+	}
+
+	l.currentSize += n
+	if l.currentSize >= l.maxSize {
+		if err = l.maybeRotate(); err != nil {
+			return 0, err
+		}
+	}
+
+	return n, nil
+}
+
+func (l *Log) maybeRotate() error {
+	// Close and remove the file instead of truncating it so that any readers can finish reading.
+	if err := l.file.Close(); err != nil {
+		return fmt.Errorf("failed to close file for rotation: %w", err)
+	}
+	if err := os.Remove(l.file.Name()); err != nil {
+		return fmt.Errorf("failed to remove file for rotation: %w", err)
+	}
+
+	file, err := os.OpenFile(l.file.Name(), os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
+	if err != nil {
+		return fmt.Errorf("failed to reopen file for rotation: %w", err)
+	}
+	l.file = file
+	l.currentSize = 0
+	return nil
+}
+
+// Close closes the log file for writing.
+func (l *Log) Close() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	defer close(l.closeCh)
+	return l.file.Close()
+}
+
+// Watch starts watching the log file for changes, returning a channel that receives updates.
+//
+// The passed context is used to stop the watcher.
+func (l *Log) Watch(ctx context.Context, follow bool) (<-chan string, error) {
+	watcher, err := tail.TailFile(l.file.Name(), tail.Config{
+		ReOpen:        true,
+		Follow:        follow,
```

Comment:
Doesn't `TailFile` require that `Follow` is true if `ReOpen` is true?

Reply:
You are right. I need to add some tests ;)